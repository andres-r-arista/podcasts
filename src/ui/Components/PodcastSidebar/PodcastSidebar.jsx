import React from "react";

import PropTypes from "prop-types";
import { useNavigate } from "react-router-dom";

export const PodcastSidebar = ({ title, author, imgUrl, id, description, className }) => {
  const navigate = useNavigate();
  return (
    <div className={`border md:m-4 md:shadow-md md:mr-8 w-1/4 ${className}`} data-testid="podcast-sidebar">
      <div className="border-b mx-2 px-8 py-4">
        <img
          src={imgUrl}
          alt={title}
          className="w-32 h-32 cursor-pointer"
          data-testid="podcast-sidebar-image"
          onClick={() => navigate(`/podcast/${id}`)}
        />
      </div>
      <div
        className="border-b mx-2 px-2 py-4 cursor-pointer"
        data-testid="podcast-sidebar-box"
        onClick={() => navigate(`/podcast/${id}`)}
      >
        <h1
          className="font-semibold truncate max-w-full text-sm px-2 pt-2"
          data-testid="podcast-sidebar-title"
          title={title}
        >
          {title}
        </h1>
        <h2
          className="truncate max-w-full text-xs px-2 pb-2"
          data-testid="podcast-sidebar-author"
          title={author}
        >{`by ${author}`}</h2>
      </div>
      <div className="md:border-b mx-2 px-2 py-4">
        <h1 className="font-semibold truncate max-w-full text-xs px-2 pt-2">
          Description:{" "}
        </h1>
        <h2 className="italic max-w-full text-xs px-2 pb-2" data-testid="podcast-sidebar-description" title={description}>
          {description}
        </h2>
      </div>
    </div>
  );
};

PodcastSidebar.propTypes = {
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  imgUrl: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  className: PropTypes.string,
};