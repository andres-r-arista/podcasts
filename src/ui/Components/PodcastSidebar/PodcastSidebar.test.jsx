import { render, screen, fireEvent } from "@testing-library/react";

import { PodcastSidebar } from "./PodcastSidebar";

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

const title = "Title";
const author = "Author";
const imgUrl = "http://www.google.com";
const id = "123";
const description = "Description";
const className = "test"

describe("Tests in <PodcastSidebar />", () => {
  test("must match the snapshot", () => {
    const { container } = render(
      <PodcastSidebar title={title} author={author} imgUrl={imgUrl} id={id} description={description} className={className} />
    );
    expect(container).toMatchSnapshot();
  });

  test("must display the correct image url", () => {
    render(
      <PodcastSidebar title={title} author={author} imgUrl={imgUrl} id={id} description={description} className={className} />
    );
    const screenImage = screen.getByTestId("podcast-sidebar-image");
    expect(screenImage.src).toContain("http://www.google.com");
  });

  test("must display the correct title", () => {
    render(
      <PodcastSidebar title={title} author={author} imgUrl={imgUrl} id={id} description={description} className={className} />
    );
    const screenTitle = screen.getByTestId("podcast-sidebar-title");
    expect(screenTitle).toHaveTextContent("Title");
  });

  test("must display the correct author", () => {
    render(
      <PodcastSidebar title={title} author={author} imgUrl={imgUrl} id={id} description={description} className={className} />
    );
    const screenAuthor = screen.getByTestId("podcast-sidebar-author");
    expect(screenAuthor).toHaveTextContent(`by ${author}`);
  });

  test("must display the correct description", () => {
    render(
      <PodcastSidebar title={title} author={author} imgUrl={imgUrl} id={id} description={description} className={className} />
    );
    const screenTitle = screen.getByTestId("podcast-sidebar-description");
    expect(screenTitle).toHaveTextContent("Description");
  });

  test("must navigate when the img is clicked", async () => {
    render(
      <PodcastSidebar title={title} author={author} imgUrl={imgUrl} id={id} description={description} className={className} />
    );
    const screenImage = screen.getByTestId("podcast-sidebar-image");
    await expect(mockedUsedNavigate).toHaveBeenCalledTimes(0);
    fireEvent.click(screenImage);
    await expect(mockedUsedNavigate).toHaveBeenCalledTimes(1);
    await expect(mockedUsedNavigate).toHaveBeenCalledWith(`/podcast/${id}`);
  });

  test("must navigate when the box is clicked", async () => {
    render(
      <PodcastSidebar title={title} author={author} imgUrl={imgUrl} id={id} description={description} className={className} />
    );
    const screenBox = screen.getByTestId("podcast-sidebar-box");
    await expect(mockedUsedNavigate).toHaveBeenCalledTimes(0);
    fireEvent.click(screenBox);
    await expect(mockedUsedNavigate).toHaveBeenCalledTimes(1);
    await expect(mockedUsedNavigate).toHaveBeenCalledWith(`/podcast/${id}`);
  });

});
