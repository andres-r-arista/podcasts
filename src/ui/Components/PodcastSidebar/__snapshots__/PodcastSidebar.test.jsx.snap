// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <PodcastSidebar /> must match the snapshot 1`] = `
<div>
  <div
    class="border md:m-4 md:shadow-md md:mr-8 w-1/4 test"
    data-testid="podcast-sidebar"
  >
    <div
      class="border-b mx-2 px-8 py-4"
    >
      <img
        alt="Title"
        class="w-32 h-32 cursor-pointer"
        data-testid="podcast-sidebar-image"
        src="http://www.google.com"
      />
    </div>
    <div
      class="border-b mx-2 px-2 py-4 cursor-pointer"
      data-testid="podcast-sidebar-box"
    >
      <h1
        class="font-semibold truncate max-w-full text-sm px-2 pt-2"
        data-testid="podcast-sidebar-title"
        title="Title"
      >
        Title
      </h1>
      <h2
        class="truncate max-w-full text-xs px-2 pb-2"
        data-testid="podcast-sidebar-author"
        title="Author"
      >
        by Author
      </h2>
    </div>
    <div
      class="md:border-b mx-2 px-2 py-4"
    >
      <h1
        class="font-semibold truncate max-w-full text-xs px-2 pt-2"
      >
        Description:
         
      </h1>
      <h2
        class="italic max-w-full text-xs px-2 pb-2"
        data-testid="podcast-sidebar-description"
        title="Description"
      >
        Description
      </h2>
    </div>
  </div>
</div>
`;
